libs/dkg/src/rbc.cpp

```diff
@@ -0,0 +1,571 @@
+}
+
+std::string RBC::MsgTypeToString(MsgType msg_type)
+{
+  switch (msg_type)
+  {
+  case MsgType::R_SEND:
+    return "r_send";
+  case MsgType::R_ECHO:
+    return "r_echo";
+  case MsgType::R_READY:
+    return "r_ready";
+  case MsgType::R_REQUEST:
+    return "r_request";
+  case MsgType::R_ANSWER:
+    return "r_answer";
+  default:
+    assert(false);
+    return "";
+  }
+}
+
+/**
+ * Creates instance of RBC
+ *
+ * @param endpoint The muddle endpoint to communicate on
+ * @param address The muddle endpoint address
+ * @param cabinet Set of muddle addresses that form RBC network
+ * @param threshold Threshold number of Byzantine peers
+ * @param dkg
+ */
+RBC::RBC(Endpoint &endpoint, MuddleAddress address, CabinetMembers const &cabinet,
+         std::size_t &threshold, DkgService &dkg_service)
+  : address_{std::move(address)}
+  , endpoint_{endpoint}
+  , current_cabinet_{cabinet}
+  , threshold_{threshold}
+  , dkg_service_{dkg_service}
+  , rbc_subscription_(endpoint.Subscribe(SERVICE_DKG, CHANNEL_BROADCAST))
+{
+  // Set subscription for rbc
+  rbc_subscription_->SetMessageHandler([this](MuddleAddress const &from, uint16_t, uint16_t,
+                                              uint16_t, muddle::Packet::Payload const &payload,
+                                              MuddleAddress) {
+    RBCSerializer serialiser(payload);
+
+    // Deserialize the RBCEnvelop
+    RBCEnvelop env;
+    serialiser >> env;
+
+    // Dispatch the event
+    OnRBC(from, env);
+  });
+}
+
+/**
+ * Resets the RBC for a new cabinet
+ */
+void RBC::ResetCabinet()
+{
+  assert(!current_cabinet_.empty());
+  assert(current_cabinet_.find(address_) != current_cabinet_.end());
+  assert(current_cabinet_.size() > 3 * threshold_);
+  std::lock_guard<std::mutex> lock{mutex_broadcast_};
+  id_ = static_cast<uint32_t>(
+      std::distance(current_cabinet_.begin(), current_cabinet_.find(address_)));
+  parties_.clear();
+  parties_.resize(current_cabinet_.size());
+  broadcasts_.clear();
+  msg_counter_ = 0;
+}
+
+/**
+ * Sends an RBCEnvelop to a particular address
+ * @param env RBCEnvelop to be serialised and sent
+ * @param address Destination muddle address
+ */
+void RBC::Send(const RBCEnvelop &env, const MuddleAddress &address)
+{
+  // Serialise the RBCEnvelop
+  RBCSerializerCounter env_counter;
+  env_counter << env;
+
+  RBCSerializer env_serializer;
+  env_serializer.Reserve(env_counter.size());
+  env_serializer << env;
+
+  endpoint_.Send(address, SERVICE_DKG, CHANNEL_BROADCAST, env_serializer.data());
+}
+
+/**
+ * Broadcasts RBCEnvelop to all subscribed to RBC channel
+ *
+ * @param env RBCEnvelop message to be serialised and broadcasted
+ */
+void RBC::Broadcast(const RBCEnvelop &env)
+{
+  // Serialise the RBCEnvelop
+  RBCSerializerCounter env_counter;
+  env_counter << env;
+
+  RBCSerializer env_serializer;
+  env_serializer.Reserve(env_counter.size());
+  env_serializer << env;
+
+  std::lock_guard<std::mutex> lock{mutex_broadcast_};
```

Comment:
don't think this lock is neccessary?

Reply:
Ok, I inserted this yesterday when I was having problems with double messages and through it might be some multithread problem.

Reply:
ah sure. Well it's not causing much harm there. Btw, you can detect multithreaded/race conditions using a tool called thread sanitizer which effectively tries to find places in your code that are unprotected. Might be a bit heavyweight for now but worth having in the back of your mind.